src/python/pants/backend/python/rules/coverage.py

```diff
@@ -0,0 +1,279 @@
+  output_pex_filename = "coverage.pex"
+  requirements_pex = await Get[Pex](
+    CreatePex(
+      output_filename=output_pex_filename,
+      requirements=PexRequirements(requirements=tuple(coverage.get_requirement_specs())),
+      interpreter_constraints=PexInterpreterConstraints(
+        constraint_set=tuple(coverage.default_interpreter_constraints)
+      ),
+      entry_point=coverage.get_entry_point(),
+      input_files_digest=plugin_file_digest,
+    )
+  )
+  return CoverageSetup(requirements_pex, output_pex_filename)
+
+
+@dataclass(frozen=True)
+class MergedCoverageData:
+  coverage_data: Digest
+
+@rule(name="Merge coverage reports")
+async def merge_coverage_reports(
+  test_results: AddressAndTestResults,
+  transitive_targets: TransitiveHydratedTargets,
+  python_setup: PythonSetup,
+  coverage_setup: CoverageSetup,
+  subprocess_encoding_environment: SubprocessEncodingEnvironment,
+) -> MergedCoverageData:
+  """Takes all python test results and merges their coverage data into a single sql file."""
+
+  coveragerc_digest = await Get[Digest](InputFilesContent, get_coveragerc_input(DEFAULT_COVERAGE_CONFIG))
+
+  coverage_directory_digests: Tuple[Digest, ...] = await MultiGet(
+    Get(
+      Digest,
+      DirectoryWithPrefixToAdd(
+        directory_digest=test_result._python_sqlite_coverage_file,
+        prefix=address.to_address().path_safe_spec,
+      )
+    )
+    for address, test_result in test_results if test_result._python_sqlite_coverage_file is not None
+  )
+
+  chrooted_sources = await Get[ChrootedPythonSources](HydratedTargets(transitive_targets.closure))
+
+  merged_input_files: Digest = await Get(
+    Digest,
+    DirectoriesToMerge(directories=(
+      *coverage_directory_digests,
+      coveragerc_digest,
+      coverage_setup.requirements_pex.directory_digest,
+      chrooted_sources.digest,
+    )),
+  )
+
+  prefixes = [f'{address.to_address().path_safe_spec}/.coverage' for address, _ in test_results]
+  coverage_args = ['combine', *prefixes]
+  request = coverage_setup.requirements_pex.create_execute_request(
+    python_setup=python_setup,
+    subprocess_encoding_environment=subprocess_encoding_environment,
+    pex_path=f'./{coverage_setup.filename}',
+    pex_args=coverage_args,
+    input_files=merged_input_files,
+    output_files=('.coverage',),
+    description=f'Merge coverage reports.',
+  )
+
+  result = await Get[ExecuteProcessResult](
+    ExecuteProcessRequest,
+    request
+  )
+  return MergedCoverageData(coverage_data=result.output_directory_digest)
+
+
+COVERAGE_PLUGIN_MODULE_NAME = '__coverage_coverage_plugin__'
+
+
+def get_coverage_plugin_input():
+  return InputFilesContent(
+    FilesContent(
+      (
+        FileContent(
+          path=f'{COVERAGE_PLUGIN_MODULE_NAME}.py',
+          content=pkg_resources.resource_string(__name__, 'coverage/plugin.py'),
+          is_executable=False,
```

Comment:
I believe we default this to `False`, so we don't need to pass the `is_executable` kwarg here.